### PR TITLE
Fully static PALs and simplified inheritance

### PR DESCRIPTION
**Pull request by nwf**
See #233 .

### CODE REVIEW

File: src/ds/aba.h [hunk 1]
```diff
@@ -85,6 +85,12 @@ namespace snmalloc
       Linked old;
       ABA* parent;
 
+      Cmp(Linked old, ABA* parent)
```
Comment (davidchisnall) [inline src/ds/aba.h:88]: Please add a comment explaining why we need this (MSVC bug?)

File: src/ds/aba.h [hunk 2]
```diff
@@ -85,6 +85,12 @@ namespace snmalloc
...
+      {
+        this->old = old;
+        this->parent = parent;
```
Comment (davidchisnall) [inline src/ds/aba.h:91]: Is there a reason this isn't using initialiser syntax?

File: src/ds/concept.h [hunk 3]
```diff
@@ -0,0 +1,21 @@
+#pragma once
+
+#ifdef __cpp_concepts
+
+#  define SNMALLOC_CONCEPT(c) c
```
Comment (davidchisnall) [inline src/ds/concept.h:5]: Doc comment.

File: src/mem/alloc.h [hunk 4]
```diff
@@ -1336,7 +1335,7 @@ namespace snmalloc
 
         slab->init(public_state(), sizeclass, rsize);
         chunkmap().set_slab(slab);
-        p = slab->alloc<zero_mem>(size, large_allocator.memory_provider);
+        p = slab->alloc<zero_mem, MemoryProvider>(size);
```
Comment (mjp41) [inline src/mem/alloc.h:1338]: Would it make more sense to take
```C++
MemoryProvider::Pal
```
as the template parameter?

File: src/mem/slab.h [hunk 5]
```diff
@@ -37,11 +37,8 @@ namespace snmalloc
      * `fast_free_list` for further allocations.
      */
     template<ZeroMem zero_mem, typename MemoryProvider>
```
Comment (mjp41) [inline src/mem/slab.h:39]: Just take the Pal as a static(template) parameter?

File: src/mem/mediumslab.h [hunk 6]
```diff
@@ -77,7 +77,7 @@ namespace snmalloc
     }
 
     template<ZeroMem zero_mem, typename MemoryProvider>
```
Comment (mjp41) [inline src/mem/mediumslab.h:79]: Take Pal as template parameter?

File: src/pal/pal_concept.h [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+#ifdef __cpp_concepts
+#  include "../ds/concept.h"
+#  include "pal_consts.h"
+
+#  include <utility>
+
+namespace snmalloc
+{
+  template<typename PAL>
```
Comment (davidchisnall) [inline src/pal/pal_concept.h:11]: Lots of doc comments here please!

**Review by mjp41 — CHANGES_REQUESTED** [review]
Could you update the documentation about writing a Pal:
https://github.com/microsoft/snmalloc#porting-snmalloc-to-a-new-platform

This needs littering with `static` now.

**Review by mjp41 — APPROVED** [review]
LGTM

**Pull request merged**